### PR TITLE
Support subqueries in non INNER join

### PR DESCRIPTION
**Pull request by kokosing**
Support subqueries in non INNER join

From now subqueries can be used in join condition for non inner join.
The only lacking support is for IN predicate (semi join) which is using
references from both tables. In such case it is kind of triple side join
which is currently not possible to be executed in Presto.

### CODE REVIEW

**Comment by ilfrin** [comment]
please add the Presto documentation changes reflecting this change (and its limitations) as part of this pull request, or better part of the relevant commit

**Comment by losipiuk** [comment]
s/non INNER/OUTER/

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java [hunk 1]
```diff
@@ -99,40 +99,29 @@ public InPredicateRewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symb
         @Override
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java]: typo i commit message "mappingse".

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java [hunk 2]
```diff
@@ -99,40 +99,29 @@ public InPredicateRewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symb
...
 
             return new FilterNode(
-                    inPredicateRewriteResult.node.getId(),
-                    getOnlyElement(inPredicateRewriteResult.node.getSources()),
+                    rewrittenNode.getId(),
+                    getOnlyElement(rewrittenNode.getSources()),
                     replaceInPredicates(node.getPredicate()));
         }
 
-        private Expression replaceInPredicates(Expression expression)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java]: if possible please do not mix:
- moving functions around the class
- with other changes.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 3]
```diff
@@ -315,15 +317,32 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
...
-                wrapWithFilterIfNeeded(rightPlanBuilder.getRoot(), preRightConditions),
+                leftPlanBuilder.getRoot(),
+                rightPlanBuilder.getRoot(),
                 equiClauses.build(),
                 Optional.empty(),
                 Optional.empty(),
                 Optional.empty());
 
+        if (node.getType() != INNER) {
+            List<InPredicate> inPredicateSubqueries = analysis.getInPredicateSubqueries(node);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: test cases when complex expression is "complex" but touches just one side of the join.
Test both outer and inner side.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 4]
```diff
@@ -315,15 +317,32 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
...
+            for (Expression complexExpression : complexJoinExpressions) {
+                if (inPredicateSubqueries.stream().anyMatch(inPredicate -> nodeContains(complexExpression, inPredicate))) {
+                    throw new UnsupportedOperationException("Unable to rewrite IN predicate with subquery in join condition");
+                }
+            }
+
+            // non inner join subqueries can be applied only to one side of join
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: does it work for subqueries touching outer symbols only?
Please add more verbose testing scenarios.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 5]
```diff
@@ -315,15 +317,32 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
...
                 equiClauses.build(),
                 Optional.empty(),
                 Optional.empty(),
                 Optional.empty());
 
+        if (node.getType() != INNER) {
+            List<InPredicate> inPredicateSubqueries = analysis.getInPredicateSubqueries(node);
+            for (Expression complexExpression : complexJoinExpressions) {
+                if (inPredicateSubqueries.stream().anyMatch(inPredicate -> nodeContains(complexExpression, inPredicate))) {
+                    throw new UnsupportedOperationException("Unable to rewrite IN predicate with subquery in join condition");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I'd rephrase as "IN with subquery predicate in join condition not supported". Also, throw a SemanticException so that the error is reported appropriately.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java [hunk 6]
```diff
@@ -121,10 +121,24 @@ public void testPlanDsl()
...
+        assertEquals(
+                countOfMatchingNodes(
+                        plan("SELECT * FROM orders WHERE orderkey = (SELECT 1) AND custkey = (SELECT 2) AND custkey != (SELECT 1)"),
+                        planNode -> planNode instanceof EnforceSingleRowNode),
+                2);
+        // same query used for left, right and complex join condition
+        assertEquals(
+                countOfMatchingNodes(
+                        plan("SELECT * FROM orders o1 JOIN orders o2 ON o1.orderkey = (SELECT 1) AND o2.orderkey = (SELECT 1) AND o1.orderkey + o2.orderkey = (SELECT 1)"),
+                        planNode -> planNode instanceof EnforceSingleRowNode),
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java]: You can do `EnforceSingleRowNode.class::isInstance` here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I don't get much the sentence: `does it work for subqueries touching outer symbols only?`. Note that it works only for non-correlated subqueries.

**Comment by kokosing** [comment]
@ilfrin @martint Rebased and most of the comments got applied. I didn't address only the doc comment as there is no documentation about joins or subqueries support at all. Documenting only this special case would be awkward, while documenting joins or subqueries support is out of the scope of this PR.

**Comment by ilfrin** [comment]
Create a ticket for this doc then, and assign yourself to it, so that doesn't fall too much behind.

**Comment by kokosing** [comment]
@martint Rebased. Ping.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 7]
```diff
@@ -2345,111 +2346,220 @@ public void testJoinWithNullValues()
...
+                ".*IN with subquery predicate in join condition not supported");
+        assertQueryFails(
+                queryPrefix.apply("left") + "y in (VALUES 1)",
+                ".*IN with subquery predicate in join condition not supported");
         assertQuery(
-                queryPrefix + "(x in (VALUES 1,2)) = (y in (VALUES 1,2)) AND (x in (VALUES 1)) != (y in (VALUES 3))",
-                "VALUES (1,2), (1,1), (3, 3)");
+                queryPrefix.apply("left") + twoDuplicatedInSubqueriesCondition,
+                "VALUES (1,1), (1,2), (2,2), (2,1), (3,3), (4, null)");
+        assertQueryFails(
```
Comment (ArturGajowy) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Please reduce testcase duplication by moving cases differing only in join type (left/right/inner/outner) into a loop. Please also ensure enough context in error messages by making assertQuery report the failing query and expected result

Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: I applied your comment, but I am not a very proud of that. I am not sure of its value.

**Comment by kokosing** [comment]
Rebased and applied comments

**Comment by kokosing** [comment]
Rebased.

@martint Can you please take a look?

**Comment by kokosing** [comment]
It seems that travis failure is related to https://github.com/prestodb/presto/issues/5854

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 8]
```diff
@@ -299,6 +301,23 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
...
+                       throw new SemanticException(
+                               NOT_SUPPORTED,
+                               inPredicate,
+                               "IN with subquery predicate in join condition not supported");
+                   }
+                }
+            }
+
+            // non inner join subqueries can be applied only to one side of join
+            leftPlanBuilder = subqueryPlanner.handleSubqueries(leftPlanBuilder, complexJoinExpressions, node);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: What's special about the left side? Is this only for left outer joins? What happens for full or right outer joins?

**Comment by martint** [comment]
I had a question above. Also, can you rebase and update the PR to see if it now passes the build?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 8 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I will rewrite the comment as it does not tell much.

There is not much special about the left side. It could be a right, not much difference. It is for any kind of outer join. 

Subquery (non LATERAL JOIN) cannot access FROM relations. `complexExpression` have to have access to all FROM relations as well as to subquery, so subquery has to be added to the one side of join. Left is selected in arbitrary way.

**Comment by kokosing** [comment]
@martint Rebased and question was answered. Can you please take a look?

**Comment by kokosing** [comment]
Superseded by #5983
